skglm/penalties/block_separable.py

```diff
@@ -324,10 +324,11 @@ def subdiff_distance(self, w, grad_ws, ws):
 
             if self.positive and np.any(w_g < 0):
                 scores[idx] = np.inf
-            if self.positive and norm_w_g == 0:
-                # distance of -norm(grad_g) to )-infty, alpha * weights[g]]
-                scores[idx] = max(0, - norm(grad_g) - self.alpha * weights[g])
-            if (not self.positive) and norm_w_g == 0:
+            elif self.positive and norm_w_g == 0:
+                # distance of -norm(neg_grad_g) to weights[g] * [-alpha, alpha]
+                neg_grad_g = np.where(grad_g < 0., grad_g, 0.)
+                scores[idx] = max(0, norm(neg_grad_g) - self.alpha * weights[g])
```

Comment:
why not using
`scores[idx] = max(0, norm(grad_g[grad_g < 0.]) - alpha * weights[g])` ?
or 
`neg_grad_g = grad_g[grad_g < 0.]) ` ?

Reply:
yes indeed @QB3, looks clear and clean

Reply:
To compute the distance to the subdifferential (for group $g$)

$$
\begin{equation}
\begin{aligned}
v \in \partial \Omega & \Leftrightarrow v \in \lambda \partial \lVert \cdot \lVert_2 + \partial \delta_{\mathbb{R}_{+}^g} \\\\
 &  \Leftrightarrow v \in \lambda \mathcal{B}_2 + \mathbb{R}^{-}_g
\end{aligned}
\end{equation}
$$

hence, the distance to subdiff is the solution of this problem (Minkowski sum)

$$
\begin{equation}
\begin{aligned}
\min_{u, n} & \ \lVert u + n - v \rVert^2 \\
\mathrm{st.} & \ u \in \lambda \\mathcal{B}_2, n \in \\mathbb{R}_g^{-}
\end{aligned}
\end{equation}
$$

The implemented formula solves this problem by solving first for $n \in \\mathbb{R}_g^{-}$, then for $u \in \lambda \\mathcal{B}_2$
Intuitively, it seems right to do that for this case, but I can't prove it rigorously

Reply:
Maybe we can extend [this proof](https://math.stackexchange.com/questions/2755045/how-is-the-subdifferential-of-the-l-2-norm-at-x-0-the-polar-of-the-unit-ball).
I would vote for merging the fix quickly, and then doing a PR to detail the subdifferential computation.

Reply:
@Badr-MOUFAD can you include the derivation in https://contrib.scikit-learn.org/skglm/tutorials/prox_nn_group_lasso.html#prox-nn-group-lasso ? Since this is not "trivial"

You're looking at a projection problem; it's clear that for the negative entries, their value in the solution is the same; then you're left with the projection of the positive entries on the ball + R_-, wich is clearly the projection on the ball too (as projecting onto R- would imply changing the sign, which is suboptimal)

Reply:
>  I would vote for merging the fix quickly, and then doing a PR to detail the subdifferential computation.

I disagree, we tend to not finish stuff that is "left for later"
The proof is not complex (we can even just give the solution)